File: arduinobot_cpp_examples/src/simple_publisher.cpp
#include <rclcpp/rclcpp.hpp>
#include <std_msgs/msg/string.hpp>

#include <chrono>
using namespace std::chrono_literals; // to use 1s

class SimplePublisher : public rclcpp::Node // inherit from Node class
{
public: 
    SimplePublisher() : Node("simple_publisher"), counter_(0) // constructor of base class
    {
        pub_ = create_publisher<std_msgs::msg::String>("chatter", 10);
        timer_ = create_wall_timer(1s, std::bind(&SimplePublisher::timerCallback, this)); // Callback defined in SimplePublisher class, use this one

        RCLCPP_INFO(get_logger(), "Publishing at 1 Hz");
    }

    void timerCallback()
    {
        auto message = std_msgs::msg::String();
        message.data = "Hello ROS 2 - counter: " + std::to_string(counter_++);
        pub_->publish(message); // on pub_ object execute publish-function
    }

private:
    unsigned int counter_;
    rclcpp::Publisher<std_msgs::msg::String>::SharedPtr pub_;
    rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char* argv[])
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<SimplePublisher>();
    rclcpp::spin(node); // run continuously
    rclcpp::shutdown(); // destructor
    return 0;
}

File: arduinobot_py_examples/arduinobot_py_examples/simple_publisher.py
import rclpy
from rclpy.node import Node # Node class from node module in rclpy library
from std_msgs.msg import String

class SimplePublisher(Node):  # inherit from Node
    def __init__(self): # constructor
        super().__init__("simple_publisher") # get input from parent class
        self.pub_ = self.create_publisher(String, "chatter",10) # use String msg to interact with topic of name chatter and buffer 10, if not receiving msg fast enough
        self.counter_ = 0 # additional variable for publisher class, count msg
        self.frequency_ = 1.0 # [Hz] msg publishing
        self.get_logger().info("Publishing at %d Hz" % self.frequency_)
        self.timer_ = self.create_timer(self.frequency_, self.timerCallback)

    def timerCallback(self):
        msg = String()
        msg.data = "Hello ROS 2 - counter: %d" % self.counter_
        self.pub_.publish(msg)
        self.counter_ += 1

def main():
    rclpy.init()
    simple_publisher = SimplePublisher()
    rclpy.spin(simple_publisher)
    simple_publisher.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()
